### DISCUSSION TITLE
[riverpod-generator] do all providers need to be refactored?

### DISCUSSION THREAD

**Discussion by samleecampspot** [rrousselGit/riverpod#2525] [Q&A] [answered]
Hello!  So I am wondering if we add the generator package, do all providers need to be refactored to use the generator?  

I was hoping to bring in the generator package, use the generator on new development, and slowly refactor my non-generated providers in bite-size pieces.  However, adding in the generator package throws a bunch of build_runner errors and I can no longer build/run my app.  Here's one of many errors:

```
../../../.pub-cache/hosted/pub.dev/riverpod-2.3.6/lib/src/async_notifier/family.dart:113:46: Error: The parameter 'create' of the method 'AsyncNotifierProviderFamily.overrideWith' has type 'NotifierT Function()', which does not match the corresponding type, 'FamilyAsyncNotifierProviderImpl<NotifierT, T, Arg> Function(Arg)', in the overridden method, 'Family with _FamilyMixin.overrideWith'.
 - 'FamilyAsyncNotifierProviderImpl' is from 'package:riverpod/src/async_notifier.dart' ('../../../.pub-cache/hosted/pub.dev/riverpod-2.3.6/lib/src/async_notifier.dart').
Change to a supertype of 'FamilyAsyncNotifierProviderImpl<NotifierT, T, Arg> Function(Arg)', or, for a covariant parameter, a subtype.
  Override overrideWith(NotifierT Function() create) {
```

If I remove the `riverpod_generator` from the pub spec, then I'm able to build/run the app successfully.

Would someone mind confirming if all providers need to be generated to use this packages?

Thanks.

**Comment by rrousselGit** [comment, marked as answer]
You should be able to slowly refactor, yes.

The error appears to be a different issue. Maybe you didn't respect the codegen syntax correctly.

**Reply by samleecampspot** [reply]
That's great!  I'll go double check my providers and see if I can get rid of these errors on my end.

Thank you Remi for the clarification and all the hard work on this package 🙏

**Reply by samleecampspot** [reply]
I do have a follow up question @rrousselGit.  

So for the `Maybe you didn't respect the codegen syntax correctly.`. Is that in regards to the new ways to setup autoDispose and family?  Like with Async notifier, there are a few different classes like `AutoDisposeAsyncNotifierProviderFamily` or `AutoDisposeAsyncNotifierProvider`?

I'm not using the generator at all, so would you mind expanding a bit on that part?

**Reply by rrousselGit** [reply]
Well technically you're never supposed to write something like "AutoDisposeAsyncNotifierProviderFamily" yourself. The codegen inputs it for you.

Make sure you added the correct `extends _$YourClass`.

If in doubt, install riverpod_lint. It warns if you forgot something.

**Reply by samleecampspot** [reply]
I believe the `extends _$YourClass` would be for generated providers only, right?  Is there any way to get around this without the generator?  I was trying to keep all my providers as close as possible to the 1.0 version when upgrading to 2.0, but I did have to manually refactor some providers to do things like:

```
final testProvider = AutoDisposeAsyncNotifierProviderFamily<SomeNotifier, Object, Object>(() {});

class SomeNotifier extends AutoDisposeFamilyAsyncNotifier<Object, Object>{}
```

I do have the riverpod_lint and custom_lint packages added, but there are no problems on my VSCode or when I run `dart run custom_lint`.  All the errors from the build runner that I'm seeing seem to complain on AutoDispose and/or Family providers for that `overrideWith` method.

It seems like I'm doing this incorrectly, but I assume doing those refactors are the cause for the errors I'm seeing on my end?

**Reply by rrousselGit** [reply]
Oh, for non-generated providers, yes, you'll have to type those AutoDisposeAsyncNotifierProviderFamily & co.

Then I'm not sure what your issue is. I don't think it's related to riverpod_generator.

**Reply by samleecampspot** [reply]
Ah okay, at least I was on the right track with the refactor.

Would you think it might have to do w/ the latest Flutter/Dart version?  I find it a bit strange to see a bunch of errors only when the riverpod_generator package is added 😕

**Reply by rrousselGit** [reply]
No. I'm pretty sure it's you extending the wrong class in a non-generated provider somehow.

**Reply by samleecampspot** [reply]
Cool!  I'll continue digging into this then.  Thanks!

**Comment by samleecampspot** [comment]
Just following up here in case someone else runs into this weird issue I had because it was definitely contained to my machine only.  CI and other devs had no issues with the build runner.

I had to run a `flutter pub cache repair` to fix the errors above.  I'm still confused why this occurred, but at least that solved it for me.
